### PR TITLE
chore: Bump Go to v1.21.0

### PR DESCRIPTION
**Pull request by codingllama**
Update Go toolchain to the latest release.

* https://go.dev/doc/go1.21

Changelog: Update Go to 1.21.0.

### CODE REVIEW

File: build.assets/Makefile [hunk 1]
```diff
@@ -20,7 +20,7 @@ OS ?= linux
 ARCH ?= amd64
 
 # Sync with devbox.json.
-GOLANG_VERSION ?= go1.20.7
+GOLANG_VERSION ?= go1.21.0
```
Comment (zmb3) [inline build.assets/Makefile:23]: Should we update devbox.json too?

**Comment by codingllama** [comment]
Building the new buildbox and verifying effects on golangci-lint right now. I've been the running the RC since it's out so I don't think we'll see any issues.

File: build.assets/Makefile (hunk 1 above)
Comment (codingllama) [reply build.assets/Makefile:23]: I doubt devbox has the new version, but I can double-check.

**Comment by mdwn** [comment]
How about `go.mod` as well?

**Comment by codingllama** [comment]
> How about `go.mod` as well?

I don't want to bump the module version, just the toolchain. There are larger effects to bumping the module itself, for example what happens with backports/existing releases.

File: build.assets/Makefile (hunk 1 above)
Comment (mdwn) [reply build.assets/Makefile:23]: Can confirm that it does not yet. The most recent RC is in nixpkgs staging: https://github.com/NixOS/nixpkgs/blob/staging/pkgs/development/compilers/go/1.21.nix.

We could potentially point to that when it's updated fully.

**Review by jakule — APPROVED** [review]

**Review by mdwn — APPROVED** [review]

**Review by rosstimothy — APPROVED** [review]

**Review by zmb3 — APPROVED** [review]

**Comment by codingllama** [comment]
e/ update just landed, so I've snuck it in as well.

**Pull request merged**